auv_navigation/auv_navigation/src/navigation_utils/path_utils.py

```diff
@@ -0,0 +1,191 @@
+        return None
+
+def calculate_carrot_pose(path, robot_pose, carrot_distance):
+    if not path.poses or carrot_distance <= 0:
+        return None
+
+    min_dist = float('inf')
+    closest_index = 0
+    
+    for i, pose in enumerate(path.poses):
+        dist = np.linalg.norm(
+            np.array([
+                pose.pose.position.x - robot_pose.pose.position.x,
+                pose.pose.position.y - robot_pose.pose.position.y,
+                pose.pose.position.z - robot_pose.pose.position.z
+            ])
+        )
+        if dist < min_dist:
+            min_dist = dist
+            closest_index = i
+
+    # If closest to last point, return it
+    if closest_index >= len(path.poses) - 1:
+        return path.poses[-1]
+    
+    # Walk along path segments until we've consumed carrot_distance
+    remaining_distance = carrot_distance
+    current_index = closest_index
+    
+    while remaining_distance > 0 and current_index < len(path.poses) - 1:
+        segment_start = path.poses[current_index].pose
+        segment_end = path.poses[current_index + 1].pose
+        
+        segment_distance = np.linalg.norm(
+            np.array([
+                segment_end.position.x - segment_start.position.x,
+                segment_end.position.y - segment_start.position.y,
+                segment_end.position.z - segment_start.position.z
+            ])
+        )
+        
+        # Skip zero-length segments
+        if segment_distance < 1e-6:
+            current_index += 1
+            continue
+
+        # If we can place carrot on this segment
+        if remaining_distance <= segment_distance:
+            # interpolation of position
+            carrot_pose = PoseStamped()
+            carrot_pose.header = path.header
+            ratio = remaining_distance / segment_distance
+            carrot_pose.pose.position.x = segment_start.position.x + \
+                ratio * (segment_end.position.x - segment_start.position.x)
+            carrot_pose.pose.position.y = segment_start.position.y + \
+                ratio * (segment_end.position.y - segment_start.position.y)
+            carrot_pose.pose.position.z = segment_start.position.z + \
+                ratio * (segment_end.position.z - segment_start.position.z)
+            
+            # Use end segment orientation as carrot orientation
+            carrot_pose.pose.orientation.x = segment_end.orientation.x
+            carrot_pose.pose.orientation.y = segment_end.orientation.y
+            carrot_pose.pose.orientation.z = segment_end.orientation.z
+            carrot_pose.pose.orientation.w = segment_end.orientation.w
+            
+            return carrot_pose
+        
+        remaining_distance -= segment_distance # Move to next segment
+        current_index += 1
+    
+    # If we've consumed all segments, return last pose
+    return path.poses[-1]
+
+
+def broadcast_carrot_frame(tf_broadcaster, tf_buffer, source_frame, carrot_pose):
```

Comment:
I think having the carrot as a frame rather than a marker is more informative for debugging. Good thing about a frame is that we can also see its orientation, which is good to know. Also, I find having a carrot frame suits more to our path following implementation for several reasons.